commands/upgrade/upgrade_path.py

```diff
@@ -0,0 +1,124 @@
+                return True
+    return False
+
+
+def get_upgrade_flavour():
+    """
+    Returns the flavour of the upgrade for this system.
+    """
+    if detect_sap_hana():
+        return LEAPP_UPGRADE_FLAVOUR_SAP_HANA
+    return LEAPP_UPGRADE_FLAVOUR_DEFAULT
+
+
+def get_os_release_version_id(filepath):
+    """
+    Retrieve data about System OS release from provided file.
+
+    :return: `str` version_id
+    """
+    with open(filepath) as f:
+        data = dict(l.strip().split('=', 1) for l in f.readlines() if '=' in l)
+        return data.get('VERSION_ID', '').strip('"')
+
+
+def get_upgrade_paths_config():
+    # NOTE(ivasilev) Importing here not to have circular dependencies
+    from leapp.cli.commands.upgrade import util  # noqa: C415; pylint: disable=import-outside-toplevel
+    from leapp.utils import path  # noqa: C415; pylint: disable=import-outside-toplevel
+
+    repository = util.load_repositories_from('repo_path', '/etc/leapp/repo.d/', manager=None)
+    with open(path.get_common_file_path(repository, LEAPP_UPGRADE_PATHS)) as f:
+        upgrade_paths_map = json.loads(f.read())
+    return upgrade_paths_map
+
+
+def get_target_versions_from_config(src_version_id, flavor):
+    """
+    Retrieve all possible target versions from upgrade_paths_map.
+    If no match is found returns empty list.
+    """
+    upgrade_paths_map = get_upgrade_paths_config()
+    return upgrade_paths_map.get(flavor, {}).get(src_version_id, [])
+
+
+def get_supported_target_versions(flavour):
+    """
+    Return a list of supported target versions for the given `flavour` of upgrade.
+    The default value for `flavour` is `default`.
+    """
+
+    current_version_id = get_os_release_version_id('/etc/os-release')
+    target_versions = get_target_versions_from_config(current_version_id, flavour)
+    if not target_versions:
+        # If we cannot find a particular major.minor version in the map,
+        # we fallback to pick a target version just based on a major version.
+        # This can happen for example when testing not yet released versions
+        major_version = get_major_version(current_version_id)
+        target_versions = get_target_versions_from_config(major_version, flavour)
+
+    return target_versions
+
+
+def get_target_version(flavour):
+    target_versions = get_supported_target_versions(flavour)
+    return target_versions[0] if target_versions else None
+
+
+def vet_upgrade_path(args):
+    """
+    Make sure the user requested upgrade_path is a supported one.
+
+    :raises: `CommandError` if the specified upgrade_path is not supported
+    :return: `tuple` (target_release, flavor)
+    """
+    flavor = get_upgrade_flavour()
+    target_release = os.getenv('LEAPP_DEVEL_TARGET_RELEASE') or args.target or get_target_version(flavor)
+    supported_target_versions = get_supported_target_versions(flavor)
+    if target_release not in supported_target_versions:
```

Comment:
I haven't checked whole code yet, but when `LEAPP_DEVEL_TARGET_RELEASE` is specified and contains data in wrong format, why should it be tested whether points to supported version? Using `LEAPP_DEVEL_TARGET_RELEASE=9.9` right now should be ok. No need for any additional check. Of course, data like `9.9.4`, `whatever`, .. should be raised as invalid.

Reply:
```
[root@localhost ~]# LEAPP_DEVEL_TARGET_RELEASE=7.1 leapp preupgrade --debug

Error: Upgrade to 7.1 for default upgrade path is not supported, possible choices are 8.4
```

In this case I would expect the warning msg in best, but upgrade should continue. Of course, it will fail in this case. but e.g. 8.3 or 8.5 should not be any problem

Reply:
I got the idea wrong then, will make LEAPP_DEVEL_TARGET_RELEASE be independent on allowed upgrade paths. Thanks

Reply:
Should be fixed now and covered by unit tests